datastore.go

```diff
@@ -392,19 +393,19 @@ func (d *Datastore) Close() error {
 
 // Batch creats a new Batch object. This provides a way to do many writes, when
 // there may be too many to fit into a single transaction.
-//
-// After writing to a Batch, always call Commit whether or not writing to the
-// batch was completed successfully or not.  This is necessary to flush any
-// remaining data and free any resources associated with an incomplete
-// transaction.
 func (d *Datastore) Batch() (ds.Batch, error) {
 	d.closeLk.RLock()
 	defer d.closeLk.RUnlock()
 	if d.closed {
 		return nil, ErrClosed
 	}
 
-	return &batch{d, d.DB.NewWriteBatch()}, nil
+	b := &batch{d, d.DB.NewWriteBatch()}
+	// Ensure that incomplete transaction resources are cleaned up in case
+	// batch is abandoned.
+	runtime.SetFinalizer(b, func(b *batch) { b.cancel() })
```

Comment:
We should probably clear the finalizer on cancel to save ourselves a bit of work. If do that, we can _also_ log an error message so we can find places in our code where we fail to call `Cancel` appropriately.

But this looks correct otherwise.

Reply:
I should have thought of that... one more small commit.